edb/pgsql/resolver/relation.py

```diff
@@ -0,0 +1,254 @@
+    )
+
+
+@dispatch._resolve_relation.register
+def resolve_relation(
+    relation: pgast.Relation, *, ctx: Context
+) -> Tuple[pgast.Relation, context.Table]:
+    assert relation.name
+    schema_name = relation.schemaname or 'public'
+
+    # try a CTE
+    cte = next((t for t in ctx.scope.ctes if t.name == relation.name), None)
+    if cte:
+        table = context.Table()
+        table.name = cte.name
+        table.columns = cte.columns.copy()
+        return pgast.Relation(name=cte.name, schemaname=None), table
+
+    # lookup the object in schema
+    obj: Optional[s_sources.Source] = None
+    if schema_name == 'public':
+        obj = ctx.schema.get(  # type: ignore
+            relation.name,
+            None,
+            module_aliases={None: 'default'},
+            type=s_objtypes.ObjectType,
+        )
+
+        # try multi link table
+        if not obj:
+            obj = _lookup_link_property(relation.name, ctx)
+
+    if not obj:
+        raise errors.QueryError(
+            f'unknown table `{schema_name}.{relation.name}`',
+            context=relation.context,
+        )
+
+    # extract table name
+    table = context.Table()
+    table.name = relation.name
+
+    # extract table columns
+    pointers = obj.get_pointers(ctx.schema).objects(ctx.schema)
+
+    columns: List[context.Column] = []
+    for p in pointers:
+        if p.is_protected_pointer(ctx.schema):
+            continue
+        card = p.get_cardinality(ctx.schema)
+        if card.is_multi():
+            continue
+
+        columns.append(_construct_column(p, ctx))
+
+    # sort by name but put `id` first
+    columns.sort(key=lambda c: () if c.name == 'id' else (c.name or '',))
+
+    table.columns.extend(columns)
+
+    # compile
+    aspect = 'inhview' if ctx.include_inherited else 'table'
+
+    schemaname, dbname = pgcommon.get_backend_name(
+        ctx.schema, obj, aspect=aspect, catenate=False
+    )
+
+    return pgast.Relation(name=dbname, schemaname=schemaname), table
+
+
+def _lookup_link_property(name: str, ctx: Context) -> Optional[s_links.Link]:
+    if '.' not in name:
+        return None
+    object_name, link_name = name.split('.')
+    parent = ctx.schema.get(  # type: ignore
+        object_name,
+        None,
+        module_aliases={None: 'default'},
+        type=s_objtypes.ObjectType,
+    )
+    if not parent:
+        return None
+
+    pointers = parent.get_pointers(ctx.schema).objects(ctx.schema)
+    for p in pointers:
+        if not isinstance(p, s_links.Link):
+            continue
+        if not p.get_shortname(ctx.schema).name == link_name:
+            continue
+
+        if p.get_cardinality(ctx.schema).is_single():
+            # single links only for tables with at least one property
+            # besides source and target
+            l_pointers = p.get_pointers(ctx.schema).objects(ctx.schema)
+            if len(l_pointers) <= 2:
+                continue
+        return p
+    return None
```

Comment:
@msullivan I'm not sure if this is the best way to do a lookup for a link of an object.

Reply:
There's a `getptr` function that is probably what you want, instead of looping through